File: lib/domain/scoring/scoring.dart
/// ============================================================================
/// SCORING SPEC v1 — the normative definition
///
/// Ch08's Cloud Function (`functions/src/submitScore.ts`, P14) re-implements
/// this in TypeScript, and the two MUST agree on every input. Treat this
/// header as the contract; change it and the port together, and bump the
/// version.
///
/// PURITY: every function below is pure — no clock, no I/O, no randomness, no
/// global state. Score depends only on its arguments. This is what lets the
/// server recompute a submission and compare.
///
/// ---------------------------------------------------------------------------
/// INPUT — an ordered list of events, replayed start to finish:
///
///   WordFound { graphemeCount: int >= 1 }   a word was correctly traced
///   WrongSelection {}                       a released selection matched nothing
///   HintUsed {}                             a paid hint was consumed
///
/// ---------------------------------------------------------------------------
/// RULES
///
/// 1. BASE POINTS
///      base = graphemeCount * 10
///    Counted in grapheme clusters, so "पानी" is 2, not 4.
///
/// 2. COMBO
///    `combo` counts CONSECUTIVE correct words. It starts at 0, increments on
///    each WordFound, and resets to 0 on WrongSelection. HintUsed does NOT
///    reset it. The multiplier is taken at the combo value AFTER incrementing,
///    capped at 6:
///
///      combo:      1     2     3     4     5     6+
///      multiplier: x1.0  x1.2  x1.4  x1.6  x1.8  x2.0
///
///    IMPLEMENTED AS INTEGER POINTS PER GRAPHEME, not as a float multiply:
///
///      combo:      1     2     3     4     5     6+
///      perGrapheme: 10    12    14    16    18    20
///
///    These are exactly `10 * multiplier`, so the ladder is identical — but a
///    float multiply is not. `0.1 + 0.2` and rounding behaviour are the classic
///    ways two languages silently disagree by one point, and one point of
///    disagreement means a rejected submission. Integers cannot diverge.
///    The TypeScript port MUST use the integer table too.
///
///      wordPoints = graphemeCount * perGrapheme[min(combo, 6)]
///
/// 3. HINT PENALTY
///      -25 points per HintUsed. Flat; not scaled by combo.
///
/// 4. FLOOR
///      The level score is clamped at 0. Hints can never push a player
///      negative.
///
/// 5. STARS — RELAXED MODE
///      3 stars = 0 hints
///      2 stars = 1 hint
///      1 star  = 2 or more hints
///
///    Stars take NO time input. That is enforced by the signature, not by a
///    convention: `computeStars` has no elapsed parameter to pass, so relaxed
///    mode cannot accidentally grow a time dependency. Timing only ever
///    matters in Blitz mode (v1.2), which will get its own function rather
///    than a flag on this one.
///
/// ---------------------------------------------------------------------------
/// WORKED EXAMPLE — used verbatim as a cross-language parity test in P14:
///
///   [WordFound(5), WordFound(4), WrongSelection, WordFound(3), HintUsed]
///
///   WordFound(5)   combo 1  ->  5 * 10 = 50    total 50
///   WordFound(4)   combo 2  ->  4 * 12 = 48    total 98
///   WrongSelection combo -> 0                  total 98
///   WordFound(3)   combo 1  ->  3 * 10 = 30    total 128
///   HintUsed                 ->      -25       total 103
///
///   computeScore = 103, computeStars(hintsUsed: 1) = 2
/// ============================================================================
library;

import 'dart:math';

import 'score_event.dart';

/// Pure scoring rules. See the spec header above.
abstract final class Scoring {
  /// Spec version. The Cloud Function checks this so a client built against
  /// older rules is detectable rather than silently mis-scored.
  static const int specVersion = 1;

  static const int basePointsPerGrapheme = 10;
  static const int hintPenalty = 25;

  /// Longest combo that still increases the multiplier.
  static const int maxComboStep = 6;

  /// Points per grapheme at combo 1..6. Integer on purpose — see rule 2.
  static const List<int> comboPointsPerGrapheme = [10, 12, 14, 16, 18, 20];

  /// The multiplier a player sees on screen ("x1.4"). Display only: scoring
  /// itself never multiplies by these, to keep Dart and TypeScript identical.
  static double comboMultiplier(int consecutiveCorrect) =>
      comboPointsPerGrapheme[_comboIndex(consecutiveCorrect)] /
      basePointsPerGrapheme;

  /// Points for one word found as the [consecutiveCorrect]-th in a row
  /// (1-based: the first correct word of a streak is 1).
  static int wordScore({
    required int graphemeCount,
    required int consecutiveCorrect,
  }) {
    if (graphemeCount <= 0) return 0;
    return graphemeCount *
        comboPointsPerGrapheme[_comboIndex(consecutiveCorrect)];
  }

  /// The level score, by replaying [events] in order.
  ///
  /// Pure. The Cloud Function runs the identical algorithm over the events the
  /// client submitted and writes ITS result, never the client's number.
  static int computeScore(List<ScoreEvent> events) {
    var score = 0;
    var combo = 0;

    for (final event in events) {
      switch (event) {
        case WordFound(:final graphemeCount):
          combo++;
          score += wordScore(
            graphemeCount: graphemeCount,
            consecutiveCorrect: combo,
          );
        case WrongSelection():
          combo = 0;
        case HintUsed():
          score -= hintPenalty;
      }
    }

    return max(score, 0);
  }

  /// Stars for a completed level in RELAXED mode.
  ///
  /// Deliberately takes no elapsed time: relaxed is the product default and
  /// timing pressure is what drives this game's audience away (Ch01/Ch02).
  static int computeStars({required int hintsUsed}) {
    if (hintsUsed <= 0) return 3;
    if (hintsUsed == 1) return 2;
    return 1;
  }

  /// Convenience for callers holding an event list rather than a counter.
  static int hintsIn(List<ScoreEvent> events) =>
      events.whereType<HintUsed>().length;

  /// The longest run of consecutive correct words in [events]. Feeds the
  /// `combo_max` analytics parameter (Ch11).
  static int maxComboIn(List<ScoreEvent> events) {
    var best = 0;
    var current = 0;
    for (final event in events) {
      switch (event) {
        case WordFound():
          current++;
          best = max(best, current);
        case WrongSelection():
          current = 0;
        case HintUsed():
          break;
      }
    }
    return best;
  }

  /// Clamps a 1-based combo position onto the ladder.
  static int _comboIndex(int consecutiveCorrect) =>
      min(max(consecutiveCorrect, 1), maxComboStep) - 1;
}

File: functions/src/scoring.ts
/**
 * ============================================================================
 * SCORING SPEC v1 — the TypeScript half of a two-language normative contract.
 *
 * THE SPEC LIVES IN `lib/domain/scoring/scoring.dart`. That file's header is
 * the normative text; this file is its port, and the two MUST agree on every
 * input. Change one without the other and `scoring_parity.test.ts` fails,
 * which is the entire reason that test exists.
 *
 * Three rules keep the two implementations from drifting apart:
 *
 *  1. INTEGER POINTS PER GRAPHEME, never a float multiply. The displayed
 *     multipliers (x1.0 … x2.0) are presentation; scoring uses
 *     `COMBO_POINTS_PER_GRAPHEME` directly. A float multiply is the classic
 *     way two languages disagree by one point, and one point of disagreement
 *     is a rejected submission.
 *  2. The same replay shape. Score is computed by walking an ORDERED event
 *     list, because the combo ladder depends on the sequence of correct and
 *     wrong selections and cannot be reconstructed from counters.
 *  3. No clock, no I/O, no randomness, no module state. Everything here is a
 *     pure function of its arguments, which is what makes "recompute the
 *     client's submission and compare" meaningful at all.
 *
 * The worked example from the Dart header is asserted verbatim in
 * `scoring.test.ts`:
 *
 *   [WordFound(5), WordFound(4), WrongSelection, WordFound(3), HintUsed]
 *     -> computeScore = 103, computeStars(1) = 2
 * ============================================================================
 */

/**
 * Spec version. `submitScore` compares this against the client's declared
 * `specVersion` so a client built against older rules is DETECTABLE rather
 * than silently mis-scored.
 */
export const SPEC_VERSION = 1;

export const BASE_POINTS_PER_GRAPHEME = 10;
export const HINT_PENALTY = 25;

/** Longest combo that still increases the multiplier. */
export const MAX_COMBO_STEP = 6;

/** Points per grapheme at combo 1..6. Integer on purpose — see rule 1. */
export const COMBO_POINTS_PER_GRAPHEME: readonly number[] = [10, 12, 14, 16, 18, 20];

/**
 * The wire discriminators from `lib/data/local/score_event_codec.dart`.
 *
 * THESE VALUES ARE FROZEN on both sides. Changing one silently invalidates
 * every outbox row queued on every device that has not synced yet; add a new
 * value instead, and teach both ends about it.
 */
export const EVENT_WORD_FOUND = 'w';
export const EVENT_WRONG_SELECTION = 'x';
export const EVENT_HINT_USED = 'h';

export type ScoreEvent =
  | { readonly t: typeof EVENT_WORD_FOUND; readonly g: number }
  | { readonly t: typeof EVENT_WRONG_SELECTION }
  | { readonly t: typeof EVENT_HINT_USED };

/** Clamps a 1-based combo position onto the ladder. Mirrors `_comboIndex`. */
function comboIndex(consecutiveCorrect: number): number {
  return Math.min(Math.max(consecutiveCorrect, 1), MAX_COMBO_STEP) - 1;
}

/**
 * The multiplier a player sees on screen ("x1.4").
 *
 * DISPLAY ONLY, exactly as in Dart: scoring never multiplies by this. It
 * exists so the two files stay line-for-line comparable — a reader checking
 * the port should not have to wonder whether a missing function means a
 * missing rule.
 */
export function comboMultiplier(consecutiveCorrect: number): number {
  return (
    COMBO_POINTS_PER_GRAPHEME[comboIndex(consecutiveCorrect)]! /
    BASE_POINTS_PER_GRAPHEME
  );
}

/**
 * Points for one word found as the [consecutiveCorrect]-th in a row (1-based:
 * the first correct word of a streak is 1).
 */
export function wordScore(graphemeCount: number, consecutiveCorrect: number): number {
  if (graphemeCount <= 0) return 0;
  return graphemeCount * COMBO_POINTS_PER_GRAPHEME[comboIndex(consecutiveCorrect)]!;
}

/**
 * The level score, by replaying [events] in order.
 *
 * This is the number that gets WRITTEN. The client's own total is never read —
 * it is not even sent (see `ScoreEventCodec`'s header).
 */
export function computeScore(events: readonly ScoreEvent[]): number {
  let score = 0;
  let combo = 0;

  for (const event of events) {
    switch (event.t) {
      case EVENT_WORD_FOUND:
        combo++;
        score += wordScore(event.g, combo);
        break;
      case EVENT_WRONG_SELECTION:
        combo = 0;
        break;
      case EVENT_HINT_USED:
        score -= HINT_PENALTY;
        break;
    }
  }

  return Math.max(score, 0);
}

/**
 * Stars for a completed level in RELAXED mode.
 *
 * Takes no elapsed time, and that is enforced by the signature on both sides:
 * relaxed mode has no clock to pass in, so it cannot accidentally grow a time
 * dependency. Blitz (v1.2) gets its own function, never a flag on this one.
 */
export function computeStars(hintsUsed: number): number {
  if (hintsUsed <= 0) return 3;
  if (hintsUsed === 1) return 2;
  return 1;
}

/** Number of `HintUsed` events. Mirrors `Scoring.hintsIn`. */
export function hintsIn(events: readonly ScoreEvent[]): number {
  return events.reduce(
    (count, event) => (event.t === EVENT_HINT_USED ? count + 1 : count),
    0,
  );
}

/** Number of `WordFound` events — the words this submission claims. */
export function wordsFoundIn(events: readonly ScoreEvent[]): number {
  return events.reduce(
    (count, event) => (event.t === EVENT_WORD_FOUND ? count + 1 : count),
    0,
  );
}

/** The longest run of consecutive correct words. Mirrors `Scoring.maxComboIn`. */
export function maxComboIn(events: readonly ScoreEvent[]): number {
  let best = 0;
  let current = 0;
  for (const event of events) {
    switch (event.t) {
      case EVENT_WORD_FOUND:
        current++;
        best = Math.max(best, current);
        break;
      case EVENT_WRONG_SELECTION:
        current = 0;
        break;
      case EVENT_HINT_USED:
        break;
    }
  }
  return best;
}
